cmd/lotus-shed/indexes.go

```diff
@@ -879,3 +904,362 @@ var backfillTxHashCmd = &cli.Command{
+			}
+
+			return nil
+		})
+
+		return g.Wait()
+	},
+}
+
+// pruneEventsIndex is a helper function that prunes the events index.db for a number of epochs starting from a specified height
+func pruneEventsIndex(_ context.Context, basePath string, startHeight, epochs int64) error {
+	dbPath := path.Join(basePath, databaseType, filter.DefaultDbFilename)
+	db, err := sql.Open(databaseDriver, dbPath+"?_txlock=immediate")
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		err := db.Close()
+		if err != nil {
+			fmt.Printf("ERROR: closing db: %s", err)
+		}
+	}()
+
+	tx, err := db.Begin()
+	if err != nil {
+		return err
+	}
+
+	var rows *sql.Rows
+	var res sql.Result
+	if epochs == 0 {
+		rows, err = tx.Query(selectEventIdsFromHeight, startHeight)
+	} else {
+		lowerBound := startHeight - epochs
+		if lowerBound < 0 {
+			lowerBound = 0
+		}
+
+		rows, err = tx.Query(selectEventIdsBetweenRange, startHeight, lowerBound)
+	}
+
+	if err != nil {
+		if err := tx.Rollback(); err != nil {
+			fmt.Printf("ERROR: rollback: %s", err)
+		}
+		return err
+	}
+
+	defer func() {
+		_ = rows.Close()
+	}()
+
+	deleteEventQuery, deleteEventEntriesQuery, deleteEventSeenQuery, err := getEventsDBDeleteQueries(rows)
+	if err != nil {
+		return xerrors.Errorf("error getting delete query: %w", err)
+	}
+
+	res, err = tx.Exec(deleteEventQuery)
+	if err != nil {
+		if err := tx.Rollback(); err != nil {
+			fmt.Printf("ERROR: rollback event: %s", err)
+		}
+		return err
+	}
+
+	rowsAffected, err := res.RowsAffected()
+	if err != nil {
+		return xerrors.Errorf("error getting event table rows affected: %s", err)
+	}
+
+	log.Infof("pruned %s table, rows affected: %d", "event", rowsAffected)
+
+	res, err = tx.Exec(deleteEventEntriesQuery)
+	if err != nil {
+		if err := tx.Rollback(); err != nil {
+			fmt.Printf("ERROR: rollback event_entries: %s", err)
+		}
+		return err
+	}
+
+	rowsAffected, err = res.RowsAffected()
+	if err != nil {
+		return xerrors.Errorf("error getting event entries table rows affected: %s", err)
+	}
+
+	log.Infof("pruned %s table, rows affected: %d", "event_entries", rowsAffected)
+
+	res, err = tx.Exec(deleteEventSeenQuery)
+	if err != nil {
+		if err := tx.Rollback(); err != nil {
+			fmt.Printf("ERROR: rollback event_seen: %s", err)
+		}
+		return err
+	}
+
+	if err := tx.Commit(); err != nil {
+		return err
+	}
+
+	rowsAffected, err = res.RowsAffected()
+	if err != nil {
+		return xerrors.Errorf("error getting event seen table rows affected: %s", err)
+	}
+
+	log.Infof("pruned %s table, rows affected: %d", "event_seen", rowsAffected)
+
+	return nil
+}
+
+func getEventsDBDeleteQueries(rows *sql.Rows) (string, string, string, error) {
+	var (
+		eventIds []int64
+		heights  []int64
+	)
+	for rows.Next() {
+		var (
+			id     int64
+			height int64
+		)
+		if err := rows.Scan(&id, &height); err != nil {
+			return "", "", "", xerrors.Errorf("error scanning event id: %w", err)
+		}
+		eventIds = append(eventIds, id)
```

Comment:
oh, this could get big, probably not ideal for an sql query

Reply:
Updated.